src/pagination/pagination.js

```diff
@@ -0,0 +1,194 @@
+/*
+Copyright (c) 2018 Uber Technologies, Inc.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
+*/
+// @flow
+/* global document */
+import * as React from 'react';
+import memoize from 'memoize-one';
+// Files
+import {Button, StyledBaseButton, KIND, SIZE} from '../button';
+import {StatefulMenu as Menu} from '../menu';
+import {
+  Root as StyledRoot,
+  MaxLabel as StyledMaxLabel,
+  DropdownContainer as StyledDropdownContainer,
+  DropdownMenu as StyledDropdownMenu,
+  DropdownButton as StyledDropdownButton,
+} from './styled-components';
+import {ArrowLeft, ArrowRight, ArrowDown} from './icons';
+import {getOverrideObject} from '../helpers/overrides';
+import type {PaginationPropsT, PaginationStateT} from './types';
+
+type MenuItemT = {
+  label: number,
+};
+
+export default class Pagination extends React.PureComponent<
+  PaginationPropsT,
+  PaginationStateT,
+> {
+  static defaultProps = {
+    prepositionLabel: 'of',
+    overrides: {},
+  };
+
+  state = {isMenuOpen: false};
+
+  dropdownContainerRef = React.createRef();
+
+  onPageClick = (event: MouseEvent) => {
+    const el: ?HTMLDivElement = this.dropdownContainerRef.current;
+    /* eslint-disable-next-line flowtype/no-weak-types */
+    if (el && !el.contains((event.target: any))) {
+      this.setState({isMenuOpen: false});
+    }
+  };
+
+  getMenuOptions = memoize(numPages => {
+    const menuOptions = [];
+    for (let i = 1; i <= numPages; i++) {
+      menuOptions.push(({label: i}: MenuItemT));
+    }
+    return menuOptions;
+  });
+
+  onMenuItemSelect = (item: MenuItemT) => {
+    const {onPageChange, currentPage} = this.props;
+    const page = item.label;
+    if (page !== currentPage) {
+      onPageChange && onPageChange(page, currentPage);
+      this.onDropdownButtonClick();
+    }
+  };
+
+  onPrevClick = (...args: *) => {
+    const {currentPage, onPageChange, onPrevClick} = this.props;
+    onPageChange && onPageChange(currentPage - 1, currentPage);
+    onPrevClick && onPrevClick(args);
+  };
+
+  onNextClick = (...args: *) => {
+    const {currentPage, onPageChange, onNextClick} = this.props;
+    onPageChange && onPageChange(currentPage + 1, currentPage);
+    onNextClick && onNextClick(args);
+  };
+
+  onDropdownButtonClick = () => {
+    const isMenuOpen = !this.state.isMenuOpen;
+    // no __BROWSER__ check because click only happens client-side
+    if (isMenuOpen) {
```

Comment:
I had that initially here too but it seemed unnecessary, especially for `Pagination` since users have other ways to change the page (via the buttons) - meaning the drop-down might only rarely be opened.

In general, are there concerns with regards to attaching DOM events only when we need it? We might need to align on a best-practice there too.

Reply:
it is more consistent to have handler always while it is needed. It is always needed while pagination component is present, not while it's part is present.

Reply:
Hm why is it always needed while the Pagination component is present? Its direct function is to close the drop-down menu, right? I would think that it's only needed when the drop-down is open. Are there cases where we need it when the drop-down isn't open?

Reply:
you rely on the `isMenuOpen` prop, which may not be synced with the real dropdown itself. If someone opens it and forgets to update this prop, your logic will not work. This is how bugs appear. We think that something will never happen if we use all properly. The logic in handler is always actual while pagination is rendered: if we click outside of this dropdown, we should see it close( even if it is probably closed)

Reply:
We should probably align on how state should be used and handled in our components. In this case, `isMenuOpen` is internal state to this component and isn't controlled in any other way other than through interacting with the components.

`Menu` itself doesn't know anything about open/close state by design. Can you elaborate more on where things might go out of sync here?

Reply:
Direct manipulation with DOM seems like unintended usage of this component in my opinion. E2E test frameworks should still go through the normal mean to activate the drop-down, no? In this case the test story would be "drop-down should open when clicking button" etc...

Reply:
Interesting, but since our drop-down is handled via component state, the only way (that I can see) to get this to display is to manipulate the state of this directly. Is that something we should optimize for (i.e. reaching in and manually changing the state of this item and then expecting a click to close it).

Perhaps the other approach here is to attach/remove these listeners according to the state changes?